aws/sdk-codegen/src/main/kotlin/software/amazon/smithy/rustsdk/endpoints/OperationInputTestGenerator.kt

```diff
@@ -0,0 +1,172 @@
+import software.amazon.smithy.rust.codegen.core.util.dq
+import software.amazon.smithy.rust.codegen.core.util.expectMember
+import software.amazon.smithy.rust.codegen.core.util.inputShape
+import software.amazon.smithy.rust.codegen.core.util.nonEmptyOrNull
+import software.amazon.smithy.rust.codegen.core.util.orNull
+import software.amazon.smithy.rust.codegen.core.util.toSnakeCase
+import java.util.logging.Logger
+
+class OperationInputTestDecorator : ClientCodegenDecorator {
+    override val name: String = "OperationInputTest"
+    override val order: Byte = 0
+
+    override fun extras(codegenContext: ClientCodegenContext, rustCrate: RustCrate) {
+        val endpointTests = EndpointTypesGenerator.fromContext(codegenContext).tests.nonEmptyOrNull() ?: return
+        rustCrate.integrationTest("endpoint_tests") {
+            Attribute(Attribute.cfg("feature = \"${TestUtilFeature.name}\"")).render(this, AttributeKind.Inner)
+            val tests = endpointTests.flatMap { test ->
+                val generator = OperationInputTestGenerator(codegenContext, test)
+                test.operationInputs.filterNot { usesUnsupportedBuiltIns(it) }.map { operationInput ->
+                    generator.generateInput(operationInput)
+                }
+            }
+            tests.join("\n")(this)
+        }
+    }
+}
+
+private val deprecatedBuiltins =
+    setOf(Builtins.S3_USE_GLOBAL_ENDPOINT, Builtins.STS_USE_GLOBAL_ENDPOINT).map { it.builtIn.get() }
+
+fun usesUnsupportedBuiltIns(testOperationInput: EndpointTestOperationInput): Boolean {
+    return testOperationInput.builtInParams.members.map { it.key.value }.any { deprecatedBuiltins.contains(it) }
+}
+
+/**
+ * Eventually, we need to pull this test into generic smithy. However, this relies on generic smithy clients
+ * supporting middleware and being instantiable from config (https://github.com/awslabs/smithy-rs/issues/2194)
+ *
+ * Doing this in AWS codegen allows us to actually integration test generated clients.
+ */
+class OperationInputTestGenerator(private val ctx: ClientCodegenContext, private val test: EndpointTestCase) {
+    private val runtimeConfig = ctx.runtimeConfig
+    private val moduleName = ctx.moduleUseName()
+    private val endpointCustomizations = ctx.rootDecorator.endpointCustomizations(ctx)
+    private val model = ctx.model
+    val symbolProvider = ctx.symbolProvider
+    private val instantiator = clientInstantiator(ctx)
+
+    private fun EndpointTestOperationInput.operationId() =
+        ShapeId.fromOptionalNamespace(ctx.serviceShape.id.namespace, operationName)
+
+    fun generateInput(testOperationInput: EndpointTestOperationInput) = writable {
+        val operationName = testOperationInput.operationName.toSnakeCase()
+        if (test.expect.endpoint.orNull()?.properties?.get("authSchemes")?.asArrayNode()?.orNull()
+            ?.map { it.expectObjectNode().expectStringMember("name").value }?.contains("sigv4a") == true
+        ) {
```

Comment:
This is complex enough that we should consider extracting it into its own function and documenting it.